In [58]:
from os import times

import pandas as pd
import numpy as np

from pandas.core.computation.ops import isnumeric

In [2]:
#Maria Oros

#------- Instructions from link --------
#First: explore the data

#Review the unstructured csv files and answer the following questions with code that supports your conclusions:

#Are there any data quality issues present?
#Are there any fields that are challenging to understand?
#We recommend using SQL or python and data visualization to examine the data.

#----- to run this code you can also do from terminal:
#python 2_processing/1_explore_data.py run

In [21]:
!pip install seaborn


[notice] A new release of pip is available: 23.2.1 -> 25.0.1
[notice] To update, run: pip install --upgrade pip


# Data Exploration step 1

In [4]:
# Reading files
products = pd.read_csv('../1_data/raw/PRODUCTS_TAKEHOME.csv', sep=',')
transaction = pd.read_csv('../1_data/raw/TRANSACTION_TAKEHOME.csv', sep=',')
user = pd.read_csv('../1_data/raw/USER_TAKEHOME.csv', sep=',')

## Understanding the data: PRODUCTS

seems like the categories repreent hirarchies of the product classification eg the product made by such manufacturer comes from health and wellness > belong to Hair care > Hair care accesories [no 4th categroy] of the product

In [6]:
products

CATEGORY_1                                CATEGORY_2  \
0       Health & Wellness                             Sexual Health   
1                  Snacks                             Puffed Snacks   
2       Health & Wellness                                 Hair Care   
3       Health & Wellness                                 Oral Care   
4       Health & Wellness                    Medicines & Treatments   
...                   ...                                       ...   
845547  Health & Wellness  Topical Muscle & Joint Relief Treatments   
845548             Snacks                                   Cookies   
845549             Snacks                                     Candy   
845550             Snacks                              Nuts & Seeds   
845551  Health & Wellness                                 First Aid   

                         CATEGORY_3 CATEGORY_4  \
0       Conductivity Gels & Lotions        NaN   
1              Cheese Curls & Puffs        NaN   
2             Hair Care Accessories        NaN   
3                        Toothpaste        NaN   
4                    Essential Oils        NaN   
...                             ...        ...   
845547               Braces & Wraps        NaN   
845548                          NaN        NaN   
845549             Confection Candy        NaN   
845550                    Hazelnuts        NaN   
845551               First Aid Kits        NaN   

                                             MANUFACTURER            BRAND  \
0                                                     NaN              NaN   
1                                                     NaN              NaN   
2                                PLACEHOLDER MANUFACTURER          ELECSOP   
3                                       COLGATE-PALMOLIVE          COLGATE   
4       MAPLE HOLISTICS AND HONEYDEW PRODUCTS INTERCHA...  MAPLE HOLISTICS   
...                                                   ...              ...   
845547                                                NaN              NaN   
845548                              TREEHOUSE FOODS, INC.        LOFTHOUSE   
845549                                HARIBO GMBH & CO KG           HARIBO   
845550                                     DOUBLE-COLA CO            JUMBO   
845551                                                 3M          NEXCARE   

             BARCODE  
0       7.964944e+11  
1       2.327801e+10  
2       4.618178e+11  
3       3.500047e+10  
4       8.068109e+11  
...              ...  
845547  7.223016e+11  
845548  4.182082e+10  
845549  1.001672e+11  
845550  7.539076e+10  
845551  7.967933e+11  

[845552 rows x 7 columns]

In [20]:
# Data Exploration step 1: Checking for missing values and data stats description
print("----------------------------------- Products")
print("\nProducts: Total ", len(products), " and Count of NaN in each column relative to the dataset total:\n", 100*products.isnull().sum()/len(products),
      "\nDescribe \n", products.describe(),
      "\nData types \n", products.dtypes)


----------------------------------- Products

Products: Total  845552  and Count of NaN in each column relative to the dataset total:
 CATEGORY_1       0.013128
CATEGORY_2       0.168411
CATEGORY_3       7.162895
CATEGORY_4      92.021898
MANUFACTURER    26.784160
BRAND           26.783923
BARCODE          0.476020
BARCODE_str      0.000000
dtype: float64 
Describe 
             BARCODE
count  8.415270e+05
mean   6.016109e+11
std    1.022530e+12
min    1.850000e+02
25%    7.124923e+10
50%    6.344185e+11
75%    7.683955e+11
max    6.291108e+13 
Data types 
 CATEGORY_1       object
CATEGORY_2       object
CATEGORY_3       object
CATEGORY_4       object
MANUFACTURER     object
BRAND            object
BARCODE         float64
BARCODE_str      object
dtype: object


In [10]:
products.columns


Index(['CATEGORY_1', 'CATEGORY_2', 'CATEGORY_3', 'CATEGORY_4', 'MANUFACTURER',
       'BRAND', 'BARCODE'],
      dtype='object')

In [12]:
products.groupby('MANUFACTURER')['BARCODE'].count()


MANUFACTURER
1-800-FLOWERS, INC.                434
1-800-SHAVED-ICE, LLC               11
1111FOURONE                          9
1440 FOODS COMPANY                 178
18.21 MAN MADE, LLC                 12
                                  ... 
ZOMICKS KOSHER KAFE & BAKE SHOP      1
ZORLU USA, INC.                     11
ZOTOS INTERNATIONAL INC              5
ZURU LLC                            12
ZWANENBERG FOOD GROUP (USA)          3
Name: BARCODE, Length: 4354, dtype: int64

In [17]:
# Number of observations in the dataset, count by barcode

products.groupby(['MANUFACTURER','CATEGORY_1','CATEGORY_2','CATEGORY_3','CATEGORY_4'])['BARCODE'].count()

MANUFACTURER         CATEGORY_1         CATEGORY_2              CATEGORY_3                               CATEGORY_4                 
1-800-FLOWERS, INC.  Snacks             Puffed Snacks           Popcorn                                  Already Popped Popcorn         224
                                                                                                         Microwave Popcorn                2
1111FOURONE          Health & Wellness  Hair Care               Hair Tools                               Hair Dryers                      2
                                        Hair Removal            Shaving Gel & Cream                      Women's Shaving Gel & Cream      1
18.21 MAN MADE, LLC  Health & Wellness  Hair Removal            Shaving Gel & Cream                      Women's Shaving Gel & Cream      2
                                                                                                                                       ... 
ZIOKOK               Health

In [18]:
products['BARCODE'] # will be better assessed as string


0         7.964944e+11
1         2.327801e+10
2         4.618178e+11
3         3.500047e+10
4         8.068109e+11
              ...     
845547    7.223016e+11
845548    4.182082e+10
845549    1.001672e+11
845550    7.539076e+10
845551    7.967933e+11
Name: BARCODE, Length: 845552, dtype: float64

In [31]:
products_clean = products[~products['BARCODE'].isna()]

print(len(products_clean))

841527


CONCLUSIONS ON THE QUESTIONS:
Are there any data quality issues present?
    YES, MISSINGS:
    From 1-4 categories, the Category 4 has 92% of the total data as missing, wonder if it is an issue on how the data is captured or nature of the business eg no standard categorization of item at that level
    Barcode has missings .047% of missings: this is relevant because bar code matches wIth Products table, in such case as as those are few, need to be excluded when merging tables.
    Manufacturer and brand are also with missing information (26% of the data): relevant if a research question is about that information

Are there any fields that are challenging to understand?
    Not sure how to interpret this question. For this table, fields were descriptive by the name and values.

## Transactions

- Products and transactions match at barcode, important to have the same data types and match them, would be better to match strings

In [34]:
transaction.head(20)

RECEIPT_ID PURCHASE_DATE  \
0   0000d256-4041-4a3e-adc4-5623fb6e0c99    2024-08-21   
1   0001455d-7a92-4a7b-a1d2-c747af1c8fd3    2024-07-20   
2   00017e0a-7851-42fb-bfab-0baa96e23586    2024-08-18   
3   000239aa-3478-453d-801e-66a82e39c8af    2024-06-18   
4   00026b4c-dfe8-49dd-b026-4c2f0fd5c6a1    2024-07-04   
5   0002d8cd-1701-4cdd-a524-b70402e2dbc0    2024-06-24   
6   000550b2-1480-4c07-950f-ff601f242152    2024-07-06   
7   00096c49-8b04-42f9-88ce-941c5e06c4a7    2024-08-19   
8   000e1d35-15e5-46c6-b6b3-33653ed3d27e    2024-08-13   
9   0010d87d-1ad2-4e5e-9a25-cec736919d15    2024-08-04   
10  00177c13-f50e-4fbe-839e-47dbe20a39f0    2024-07-15   
11  0019ec79-cbb3-41ed-b84c-cd74d04553f8    2024-06-25   
12  001e5563-cdec-4d46-8493-0d118a55b14c    2024-09-01   
13  001f2f3f-1746-4217-a98f-73c63c63bae2    2024-08-20   
14  0024382b-35e6-43cc-b698-133a8f192e08    2024-06-24   
15  0028d0d0-f8c5-4160-8ee6-e5b089167788    2024-08-27   
16  002bc61e-2bbf-4f4e-a3b0-d7b6cbcf9bc1    2024-07-11   
17  002ee298-d907-40ca-921a-556468571f76    2024-07-15   
18  00326689-e763-4b27-9ad5-202fc93609e2    2024-06-19   
19  00336a75-8219-4192-9c9e-c7f7a999a36d    2024-07-27   

                    SCAN_DATE              STORE_NAME  \
0   2024-08-21 14:19:06.539 Z                 WALMART   
1   2024-07-20 09:50:24.206 Z                    ALDI   
2   2024-08-19 15:38:56.813 Z                 WALMART   
3   2024-06-19 11:03:37.468 Z               FOOD LION   
4   2024-07-05 15:56:43.549 Z                RANDALLS   
5   2024-06-24 19:44:54.247 Z                 WALMART   
6   2024-07-06 19:27:48.586 Z                 WALMART   
7   2024-08-21 17:35:21.902 Z                  TARGET   
8   2024-08-13 18:21:07.931 Z                 WALMART   
9   2024-08-04 18:01:47.787 Z                    ALDI   
10  2024-07-16 09:51:06.183 Z                 WALMART   
11  2024-06-25 13:15:31.270 Z                  COSTCO   
12  2024-09-03 09:58:58.757 Z                 WALMART   
13  2024-08-22 12:53:37.811 Z                 WALMART   
14  2024-06-25 09:46:15.389 Z  DOLLAR TREE STORES INC   
15  2024-08-27 17:18:23.675 Z                 WALMART   
16  2024-07-13 20:03:24.809 Z                 WALMART   
17  2024-07-16 16:42:19.211 Z                    ALDI   
18  2024-06-20 08:59:38.397 Z                    ALDI   
19  2024-08-07 12:14:36.833 Z           FAMILY DOLLAR   

                     USER_ID       BARCODE FINAL_QUANTITY FINAL_SALE  
0   63b73a7f3d310dceeabd4758  1.530001e+10           1.00             
1   62c08877baa38d1a1f6c211a           NaN           zero       1.49  
2   60842f207ac8b7729e472020  7.874223e+10           1.00             
3   63fcd7cea4f8442c3386b589  7.833997e+11           zero       3.49  
4   6193231ae9b3d75037b0f928  4.790050e+10           1.00             
5   5dcc6c510040a012b8e76924  6.811314e+11           zero       1.46  
6   5f850bc9cf9431165f3ac175  4.920091e+10           1.00             
7   6144f4f1f3ef696919f54b5c  7.830007e+10           zero       3.59  
8   61a6d926f998e47aad33db66  5.200001e+10           1.00             
9   66686fc2e04f743a096ea808           NaN           zero       2.29  
10  61e962f64e54c329499f2df2  7.874235e+10           1.00             
11  5c62145dba25631369c28f32  5.150025e+10           zero      10.99  
12  5b5cc7cfd1c6b436c0d93abd  3.800014e+10           1.00             
13  5f57cc80a2b29216020fc68d  4.600029e+10           zero       0.97  
14  63cf692c3d310dceead1af91  2.840041e+10           1.00             
15  5ec16e6fd474d413da264f52  7.876514e+11           zero       7.48  
16  667c65a986cc41b000cd5a42  7.294561e+10           1.00             
17  63de64b1dcb50fbd3084f142           NaN           zero       2.49  
18  6158642597d737581b5d30ee           NaN           1.00             
19  665dc8fae04f743a096d1810  4.203710e+10           zero       5.25

In [17]:
print("----------------------------------- Transaction")
print("\nTransaction: Total ", len(transaction), " and Count of NaN in each column relative to the dataset total:\n", 100*transaction.isnull().sum()/len(transaction), "\nDescribe \n", transaction.describe(), "\nData types \n", transaction.dtypes)

----------------------------------- Transaction

Transaction: Total  50000  and Count of NaN in each column relative to the dataset total:
 RECEIPT_ID         0.000
PURCHASE_DATE      0.000
SCAN_DATE          0.000
STORE_NAME         0.000
USER_ID            0.000
BARCODE           11.524
FINAL_QUANTITY     0.000
FINAL_SALE         0.000
dtype: float64 
Describe 
             BARCODE
count  4.423800e+04
mean   1.715863e+11
std    3.269219e+11
min   -1.000000e+00
25%    3.077212e+10
50%    5.210004e+10
75%    8.536765e+10
max    9.347108e+12 
Data types 
 RECEIPT_ID         object
PURCHASE_DATE      object
SCAN_DATE          object
STORE_NAME         object
USER_ID            object
BARCODE           float64
FINAL_QUANTITY     object
FINAL_SALE         object
dtype: object


Some fields as dates are not date types, need to treat properly. FINAL_SALE and FINAL_NUMERIC are both character types, but the entity relation model indicate should numeric. FINAL_SALE is missing data equal to zero?


In [35]:
transaction.groupby('FINAL_QUANTITY')['RECEIPT_ID'].count() #some quantities are less than 1 eg .7, .9, does not make sense

FINAL_QUANTITY
0.01        1
0.04        1
0.09        2
0.23        4
0.24        1
        ...  
6.22        1
7.00        7
8.00        8
9.00        3
zero    12500
Name: RECEIPT_ID, Length: 87, dtype: int64

In [36]:
transaction.groupby(['FINAL_QUANTITY','FINAL_SALE'])['RECEIPT_ID'].count()


FINAL_QUANTITY  FINAL_SALE
0.01            0.05           1
0.04            0.24           1
0.09                           1
                0.12           1
0.23                           1
                              ..
zero            9.96           9
                9.97          30
                9.98          35
                9.99          68
                90.00          1
Name: RECEIPT_ID, Length: 3141, dtype: int64

In [43]:
print(transaction)

                                 RECEIPT_ID PURCHASE_DATE  \
0      0000d256-4041-4a3e-adc4-5623fb6e0c99    2024-08-21   
1      0001455d-7a92-4a7b-a1d2-c747af1c8fd3    2024-07-20   
2      00017e0a-7851-42fb-bfab-0baa96e23586    2024-08-18   
3      000239aa-3478-453d-801e-66a82e39c8af    2024-06-18   
4      00026b4c-dfe8-49dd-b026-4c2f0fd5c6a1    2024-07-04   
...                                     ...           ...   
49995  b5cd61a9-8033-4913-a5c4-fb3f65e3a321    2024-08-21   
49996  e1b2f634-c9ad-4152-b662-4b22efc25862    2024-08-11   
49997  b07ef8dd-e444-40a2-819b-f74a3e5f1ae7    2024-07-11   
49998  42475141-bef4-4df2-aa37-72577e2512bb    2024-06-18   
49999  3a179c4e-46f2-4126-b3d2-3514afc23a3e    2024-08-07   

                       SCAN_DATE     STORE_NAME                   USER_ID  \
0      2024-08-21 14:19:06.539 Z        WALMART  63b73a7f3d310dceeabd4758   
1      2024-07-20 09:50:24.206 Z           ALDI  62c08877baa38d1a1f6c211a   
2      2024-08-19 15:38:56.813 Z    

In [95]:
transaction['FINAL_QUANTITY_NUMERIC'] = transaction['FINAL_QUANTITY'].apply(lambda x: float(x) if x != 'zero' else 0.0)

transaction_clean = transaction[
    (transaction['FINAL_QUANTITY_NUMERIC'] >= 1.0) | (transaction['FINAL_QUANTITY_NUMERIC'] <= 0.0)]

print(len(transaction_clean))



49968


In [96]:
# sale numeric

transaction_clean['FINAL_SALE_NUMERIC'] = transaction_clean['FINAL_SALE'].apply(
    lambda x: 0.0 if pd.isna(x) or x == '' else float(x) if str(x).strip().replace('.', '', 1).isdigit() else np.nan
)
transaction_clean[['FINAL_SALE','FINAL_SALE_NUMERIC']]

FINAL_SALE  FINAL_SALE_NUMERIC
0                                NaN
1           1.49                1.49
2                                NaN
3           3.49                3.49
4                                NaN
...          ...                 ...
49995       1.18                1.18
49996       2.00                2.00
49997      20.96               20.96
49998       3.00                3.00
49999       5.48                5.48

[49968 rows x 2 columns]

In [98]:
transaction_clean[transaction_clean['FINAL_SALE_NUMERIC'].isna()]['FINAL_QUANTITY_NUMERIC'].describe()

count    12492.000000
mean         1.090929
std          2.498554
min          1.000000
25%          1.000000
50%          1.000000
75%          1.000000
max        276.000000
Name: FINAL_QUANTITY_NUMERIC, dtype: float64

In [ ]:
transaction_clean['FINAL_SALE_NUMERIC']=transaction_clean['FINAL_SALE_NUMERIC'].fillna(0)


In [93]:
transaction_clean

RECEIPT_ID PURCHASE_DATE  \
0      0000d256-4041-4a3e-adc4-5623fb6e0c99    2024-08-21   
1      0001455d-7a92-4a7b-a1d2-c747af1c8fd3    2024-07-20   
2      00017e0a-7851-42fb-bfab-0baa96e23586    2024-08-18   
3      000239aa-3478-453d-801e-66a82e39c8af    2024-06-18   
4      00026b4c-dfe8-49dd-b026-4c2f0fd5c6a1    2024-07-04   
...                                     ...           ...   
49995  b5cd61a9-8033-4913-a5c4-fb3f65e3a321    2024-08-21   
49996  e1b2f634-c9ad-4152-b662-4b22efc25862    2024-08-11   
49997  b07ef8dd-e444-40a2-819b-f74a3e5f1ae7    2024-07-11   
49998  42475141-bef4-4df2-aa37-72577e2512bb    2024-06-18   
49999  3a179c4e-46f2-4126-b3d2-3514afc23a3e    2024-08-07   

                       SCAN_DATE     STORE_NAME                   USER_ID  \
0      2024-08-21 14:19:06.539 Z        WALMART  63b73a7f3d310dceeabd4758   
1      2024-07-20 09:50:24.206 Z           ALDI  62c08877baa38d1a1f6c211a   
2      2024-08-19 15:38:56.813 Z        WALMART  60842f207ac8b7729e472020   
3      2024-06-19 11:03:37.468 Z      FOOD LION  63fcd7cea4f8442c3386b589   
4      2024-07-05 15:56:43.549 Z       RANDALLS  6193231ae9b3d75037b0f928   
...                          ...            ...                       ...   
49995  2024-08-31 14:13:08.634 Z         TARGET  6154bcf098f885648de2f299   
49996  2024-08-11 18:15:56.736 Z    STOP & SHOP  60aa809f188b926b2244c974   
49997  2024-07-11 08:03:25.816 Z        WALMART  60bd26e83dc3b13a15c5f4e7   
49998  2024-06-18 19:57:32.211 Z  MARKET BASKET  6169912fac47744405af62b7   
49999  2024-08-07 15:30:07.911 Z        WALMART  64e94d64ca929250373ef6e1   

            BARCODE FINAL_QUANTITY FINAL_SALE  FINAL_QUANTITY_NUMERIC  \
0      1.530001e+10           1.00                                1.0   
1               NaN           zero       1.49                     0.0   
2      7.874223e+10           1.00                                1.0   
3      7.833997e+11           zero       3.49                     0.0   
4      4.790050e+10           1.00                                1.0   
...             ...            ...        ...                     ...   
49995  8.523911e+10           2.00       1.18                     2.0   
49996  4.610040e+10           1.00       2.00                     1.0   
49997  6.466300e+11           1.00      20.96                     1.0   
49998  4.180050e+10           1.00       3.00                     1.0   
49999  3.076607e+11           1.00       5.48                     1.0   

       FINAL_SALE_NUMERIC  
0                    0.00  
1                    1.49  
2                    0.00  
3                    3.49  
4                    0.00  
...                   ...  
49995                1.18  
49996                2.00  
49997               20.96  
49998                3.00  
49999                5.48  

[49968 rows x 10 columns]

In [94]:
## dates
pd.set_option('copy_on_write', True)
transaction_clean['PURCHASE_DATE'] = pd.to_datetime(transaction_clean['PURCHASE_DATE'])
#pd.to_datetime(transaction_clean['PURCHASE_DATE'], format='%d%b%Y:%H:%M:%S.%f')
transaction_clean["SCAN_DATE"] = pd.to_datetime(transaction_clean['SCAN_DATE'])
transaction_clean
#transaction_clean['PURCHASE_DATE_DATE']=transaction_clean['PURCHASE_DATE'].as.times()

RECEIPT_ID PURCHASE_DATE  \
0      0000d256-4041-4a3e-adc4-5623fb6e0c99    2024-08-21   
1      0001455d-7a92-4a7b-a1d2-c747af1c8fd3    2024-07-20   
2      00017e0a-7851-42fb-bfab-0baa96e23586    2024-08-18   
3      000239aa-3478-453d-801e-66a82e39c8af    2024-06-18   
4      00026b4c-dfe8-49dd-b026-4c2f0fd5c6a1    2024-07-04   
...                                     ...           ...   
49995  b5cd61a9-8033-4913-a5c4-fb3f65e3a321    2024-08-21   
49996  e1b2f634-c9ad-4152-b662-4b22efc25862    2024-08-11   
49997  b07ef8dd-e444-40a2-819b-f74a3e5f1ae7    2024-07-11   
49998  42475141-bef4-4df2-aa37-72577e2512bb    2024-06-18   
49999  3a179c4e-46f2-4126-b3d2-3514afc23a3e    2024-08-07   

                             SCAN_DATE     STORE_NAME  \
0     2024-08-21 14:19:06.539000+00:00        WALMART   
1     2024-07-20 09:50:24.206000+00:00           ALDI   
2     2024-08-19 15:38:56.813000+00:00        WALMART   
3     2024-06-19 11:03:37.468000+00:00      FOOD LION   
4     2024-07-05 15:56:43.549000+00:00       RANDALLS   
...                                ...            ...   
49995 2024-08-31 14:13:08.634000+00:00         TARGET   
49996 2024-08-11 18:15:56.736000+00:00    STOP & SHOP   
49997 2024-07-11 08:03:25.816000+00:00        WALMART   
49998 2024-06-18 19:57:32.211000+00:00  MARKET BASKET   
49999 2024-08-07 15:30:07.911000+00:00        WALMART   

                        USER_ID       BARCODE FINAL_QUANTITY FINAL_SALE  \
0      63b73a7f3d310dceeabd4758  1.530001e+10           1.00              
1      62c08877baa38d1a1f6c211a           NaN           zero       1.49   
2      60842f207ac8b7729e472020  7.874223e+10           1.00              
3      63fcd7cea4f8442c3386b589  7.833997e+11           zero       3.49   
4      6193231ae9b3d75037b0f928  4.790050e+10           1.00              
...                         ...           ...            ...        ...   
49995  6154bcf098f885648de2f299  8.523911e+10           2.00       1.18   
49996  60aa809f188b926b2244c974  4.610040e+10           1.00       2.00   
49997  60bd26e83dc3b13a15c5f4e7  6.466300e+11           1.00      20.96   
49998  6169912fac47744405af62b7  4.180050e+10           1.00       3.00   
49999  64e94d64ca929250373ef6e1  3.076607e+11           1.00       5.48   

       FINAL_QUANTITY_NUMERIC  FINAL_SALE_NUMERIC  
0                         1.0                0.00  
1                         0.0                1.49  
2                         1.0                0.00  
3                         0.0                3.49  
4                         1.0                0.00  
...                       ...                 ...  
49995                     2.0                1.18  
49996                     1.0                2.00  
49997                     1.0               20.96  
49998                     1.0                3.00  
49999                     1.0                5.48  

[49968 rows x 10 columns]

CONCLUSIONS ON THE QUESTIONS:
* Are there any data quality issues present?
    - YES, MISSINGS:
    seems to be the data set with more complete data, the only field with missing information is BARCODE about 11% onf the total entries
    - DATA TYPES:
    FINAL_QUANTITY and FINAL_SALE are not numeric but combining strings to represent zero, final quantity reported fractions of number eg .6 which does not makes much sense at frst instace, there were quantities=1 with a final sale missing, seems like an inconsistency.

* Are there any fields that are challenging to understand?
    - Fields were very descriptive them-selves, but some values did not make sense as the FINAL_QUANTITY reporting fractions lower than 1. wondering f that is the case of poor collection fo data, data processing.

## User

In [106]:
user

ID               CREATED_DATE  \
0      5ef3b4f17053ab141787697d  2020-06-24 20:17:54.000 Z   
1      5ff220d383fcfc12622b96bc  2021-01-03 19:53:55.000 Z   
2      6477950aa55bb77a0e27ee10  2023-05-31 18:42:18.000 Z   
3      658a306e99b40f103b63ccf8  2023-12-26 01:46:22.000 Z   
4      653cf5d6a225ea102b7ecdc2  2023-10-28 11:51:50.000 Z   
...                         ...                        ...   
99995  61fc06d41febf771966da8fa  2022-02-03 16:46:12.000 Z   
99996  6391e7ef90ad5449ec5f782d  2022-12-08 13:34:39.000 Z   
99997  637d5efdd6f2a49c49934dcb  2022-11-22 23:45:05.000 Z   
99998  5f0de23b05d8a6147dc0cafa  2020-07-14 16:50:04.000 Z   
99999  5e7ab436905e9512ff44abc8  2020-03-25 01:30:30.000 Z   

                      BIRTH_DATE STATE LANGUAGE  GENDER  
0      2000-08-11 00:00:00.000 Z    CA   es-419  female  
1      2001-09-24 04:00:00.000 Z    PA       en  female  
2      1994-10-28 00:00:00.000 Z    FL   es-419  female  
3                            NaN    NC       en     NaN  
4      1972-03-19 00:00:00.000 Z    PA       en  female  
...                          ...   ...      ...     ...  
99995  1992-03-16 08:00:00.000 Z    CA       en  female  
99996  1993-09-23 05:00:00.000 Z    MO       en  female  
99997  1983-04-19 00:00:00.000 Z    RI       en  female  
99998  1995-06-09 04:00:00.000 Z    DE       en  female  
99999  1995-12-15 05:00:00.000 Z    VA      NaN  female  

[100000 rows x 6 columns]

In [107]:
user.columns

Index(['ID', 'CREATED_DATE', 'BIRTH_DATE', 'STATE', 'LANGUAGE', 'GENDER'], dtype='object')

In [108]:
print("----------------------------------- User")
print("\nUser: Total ", len(user), " and Count of NaN in each column relative to the dataset total:\n", 100*user.isnull().sum()/len(user), "\nDescribe \n", user.describe(), "\nData types \n", user.dtypes)

----------------------------------- User

User: Total  100000  and Count of NaN in each column relative to the dataset total:
 ID               0.000
CREATED_DATE     0.000
BIRTH_DATE       3.675
STATE            4.812
LANGUAGE        30.508
GENDER           5.892
dtype: float64 
Describe 
                               ID               CREATED_DATE  \
count                     100000                     100000   
unique                    100000                      99942   
top     5ef3b4f17053ab141787697d  2023-01-12 18:30:15.000 Z   
freq                           1                          2   

                       BIRTH_DATE  STATE LANGUAGE  GENDER  
count                       96325  95188    69492   94108  
unique                      54721     52        2      11  
top     1970-01-01 00:00:00.000 Z     TX       en  female  
freq                         1272   9028    63403   64240   
Data types 
 ID              object
CREATED_DATE    object
BIRTH_DATE      object
STATE    

In [115]:
user_clean = user
user_clean['CREATED_DATE'] = pd.to_datetime(user_clean['CREATED_DATE'])


In [117]:
user_clean[['CREATED_DATE']]

CREATED_DATE
0     2020-06-24 20:17:54+00:00
1     2021-01-03 19:53:55+00:00
2     2023-05-31 18:42:18+00:00
3     2023-12-26 01:46:22+00:00
4     2023-10-28 11:51:50+00:00
...                         ...
99995 2022-02-03 16:46:12+00:00
99996 2022-12-08 13:34:39+00:00
99997 2022-11-22 23:45:05+00:00
99998 2020-07-14 16:50:04+00:00
99999 2020-03-25 01:30:30+00:00

[100000 rows x 1 columns]

CONCLUSIONS
ON
THE
QUESTIONS:
* Are there any data quality issues present?
    - YES, MISSINGS: mostly on the field called LANGUAGE, and other socio demographics. Important that ID and CREATED_DATE are not missings
    - DATA TYPES: CREATED_DATE and BIRTH_DATE are characters

* Are there any fields that are challenging to understand?
    - Fields were very descriptive

## Merging data

In [122]:
transaction_clean['USER_ID'] = transaction_clean['USER_ID'].astype(str)

user_clean['ID'] = user_clean['ID'].astype(str)


In [156]:
## testing inner join to count how many users id match the transactions
users_transactions_inner = user_clean.merge(
    transaction_clean,
    left_on='ID',
    right_on='USER_ID',
    how='inner'
)
len(users_transactions_inner) #just few, why? users that does not make transactions?

262

In [167]:
pwd

'/Users/mariaoros/Documents/Github-projects/2025/Fetch-DataAnalystHome/2_processing'

In [168]:
users_transactions_inner.to_csv('../1_data/processed/users_transactions_inner.csv')

In [158]:
users_transactions = user_clean.merge(
    transaction_clean,
    left_on='ID',
    right_on='USER_ID',
    how='left'
)
(users_transactions)

ID              CREATED_DATE  \
0       5ef3b4f17053ab141787697d 2020-06-24 20:17:54+00:00   
1       5ff220d383fcfc12622b96bc 2021-01-03 19:53:55+00:00   
2       6477950aa55bb77a0e27ee10 2023-05-31 18:42:18+00:00   
3       658a306e99b40f103b63ccf8 2023-12-26 01:46:22+00:00   
4       653cf5d6a225ea102b7ecdc2 2023-10-28 11:51:50+00:00   
...                          ...                       ...   
100166  61fc06d41febf771966da8fa 2022-02-03 16:46:12+00:00   
100167  6391e7ef90ad5449ec5f782d 2022-12-08 13:34:39+00:00   
100168  637d5efdd6f2a49c49934dcb 2022-11-22 23:45:05+00:00   
100169  5f0de23b05d8a6147dc0cafa 2020-07-14 16:50:04+00:00   
100170  5e7ab436905e9512ff44abc8 2020-03-25 01:30:30+00:00   

                       BIRTH_DATE STATE LANGUAGE  GENDER RECEIPT_ID  \
0       2000-08-11 00:00:00.000 Z    CA   es-419  female        NaN   
1       2001-09-24 04:00:00.000 Z    PA       en  female        NaN   
2       1994-10-28 00:00:00.000 Z    FL   es-419  female        NaN   
3                             NaN    NC       en     NaN        NaN   
4       1972-03-19 00:00:00.000 Z    PA       en  female        NaN   
...                           ...   ...      ...     ...        ...   
100166  1992-03-16 08:00:00.000 Z    CA       en  female        NaN   
100167  1993-09-23 05:00:00.000 Z    MO       en  female        NaN   
100168  1983-04-19 00:00:00.000 Z    RI       en  female        NaN   
100169  1995-06-09 04:00:00.000 Z    DE       en  female        NaN   
100170  1995-12-15 05:00:00.000 Z    VA      NaN  female        NaN   

       PURCHASE_DATE SCAN_DATE STORE_NAME USER_ID  BARCODE FINAL_QUANTITY  \
0                NaN       NaN        NaN     NaN      NaN            NaN   
1                NaN       NaN        NaN     NaN      NaN            NaN   
2                NaN       NaN        NaN     NaN      NaN            NaN   
3                NaN       NaN        NaN     NaN      NaN            NaN   
4                NaN       NaN        NaN     NaN      NaN            NaN   
...              ...       ...        ...     ...      ...            ...   
100166           NaN       NaN        NaN     NaN      NaN            NaN   
100167           NaN       NaN        NaN     NaN      NaN            NaN   
100168           NaN       NaN        NaN     NaN      NaN            NaN   
100169           NaN       NaN        NaN     NaN      NaN            NaN   
100170           NaN       NaN        NaN     NaN      NaN            NaN   

       FINAL_SALE  FINAL_QUANTITY_NUMERIC  FINAL_SALE_NUMERIC BARCODE_str  
0             NaN                     NaN                 NaN         NaN  
1             NaN                     NaN                 NaN         NaN  
2             NaN                     NaN                 NaN         NaN  
3             NaN                     NaN                 NaN         NaN  
4             NaN                     NaN                 NaN         NaN  
...           ...                     ...                 ...         ...  
100166        NaN                     NaN                 NaN         NaN  
100167        NaN                     NaN                 NaN         NaN  
100168        NaN                     NaN                 NaN         NaN  
100169        NaN                     NaN                 NaN         NaN  
100170        NaN                     NaN                 NaN         NaN  

[100171 rows x 17 columns]

In [132]:
len(users_transactions), len(transaction_clean), len(user_clean)

(100171, 49968, 100000)

In [136]:
products_clean['BARCODE_str'] = products_clean['BARCODE'].astype(str)

transaction_clean['BARCODE_str'] = transaction_clean['BARCODE'].astype(str)

In [159]:
products_transactions_inner = products_clean.merge(
    transaction_clean,
    left_on='BARCODE_str',
    right_on='BARCODE_str',
    how='inner'
)
len(products_transactions_inner) #few

24849

In [169]:
products_transactions_inner.to_csv('../1_data/processed/products_transactions_inner.csv')

In [161]:
products_transactions_inner

CATEGORY_1    CATEGORY_2             CATEGORY_3 CATEGORY_4  \
0      Health & Wellness     Hair Care  Hair Styling Products        NaN   
1      Health & Wellness     Hair Care  Hair Styling Products        NaN   
2      Health & Wellness     First Aid       Bandages & Gauze        NaN   
3      Health & Wellness     First Aid       Bandages & Gauze        NaN   
4                 Snacks       Cookies                    NaN        NaN   
...                  ...           ...                    ...        ...   
24844             Snacks  Nuts & Seeds                Peanuts        NaN   
24845             Snacks  Nuts & Seeds                Peanuts        NaN   
24846             Snacks  Nuts & Seeds                Peanuts        NaN   
24847             Snacks         Chips           Potato Chips        NaN   
24848             Snacks         Chips           Potato Chips        NaN   

                 MANUFACTURER         BRAND     BARCODE_x     BARCODE_str  \
0            PROCTER & GAMBLE        AUSSIE  3.815192e+11  381519186967.0   
1            PROCTER & GAMBLE        AUSSIE  3.815192e+11  381519186967.0   
2                          3M       NEXCARE  5.113120e+10   51131197732.0   
3                          3M       NEXCARE  5.113120e+10   51131197732.0   
4      CHATTANOOGA BAKERY INC      MOON PIE  7.210816e+10   72108161037.0   
...                       ...           ...           ...             ...   
24844      TARGET CORPORATION  FAVORITE DAY  8.523916e+10   85239156841.0   
24845      TARGET CORPORATION  FAVORITE DAY  8.523916e+10   85239156841.0   
24846      TARGET CORPORATION  FAVORITE DAY  8.523916e+10   85239156841.0   
24847            WALMART INC.   GREAT VALUE  7.874230e+10   78742295206.0   
24848            WALMART INC.   GREAT VALUE  7.874230e+10   78742295206.0   

                                 RECEIPT_ID PURCHASE_DATE  \
0      019f46f2-f05f-4005-bd34-3190d068805c    2024-09-07   
1      019f46f2-f05f-4005-bd34-3190d068805c    2024-09-07   
2      da2a744e-5f5e-42cb-ac3c-c7c613523fe1    2024-06-30   
3      da2a744e-5f5e-42cb-ac3c-c7c613523fe1    2024-06-30   
4      b2bbbd72-1b59-4a38-9d21-b4eab7c02e06    2024-08-25   
...                                     ...           ...   
24844  52d371d6-a84e-44c2-92b3-90271c26a68a    2024-07-23   
24845  52d371d6-a84e-44c2-92b3-90271c26a68a    2024-07-23   
24846  02bb446d-30ef-4017-bd2c-3979140d09c7    2024-07-30   
24847  ad59a567-17f6-494e-a4e3-7c2e4f2699f8    2024-06-15   
24848  ad59a567-17f6-494e-a4e3-7c2e4f2699f8    2024-06-15   

                       SCAN_DATE              STORE_NAME  \
0      2024-09-07 15:39:02.804 Z                 WALMART   
1      2024-09-07 15:39:02.804 Z                 WALMART   
2      2024-07-01 10:51:25.976 Z               WALGREENS   
3      2024-07-01 10:51:25.976 Z               WALGREENS   
4      2024-08-26 09:54:58.533 Z  DOLLAR TREE STORES INC   
...                          ...                     ...   
24844  2024-07-24 10:47:29.812 Z                  TARGET   
24845  2024-07-24 10:47:29.812 Z                  TARGET   
24846  2024-07-30 14:02:45.117 Z                  TARGET   
24847  2024-06-15 15:45:02.758 Z                 WALMART   
24848  2024-06-15 15:45:02.758 Z                 WALMART   

                        USER_ID     BARCODE_y FINAL_QUANTITY FINAL_SALE  \
0      5e3da3a8f8f61412f02db496  3.815192e+11           zero       4.97   
1      5e3da3a8f8f61412f02db496  3.815192e+11           1.00       4.97   
2      5f9758e056b9b21677914f8f  5.113120e+10           1.00              
3      5f9758e056b9b21677914f8f  5.113120e+10           1.00      12.49   
4      626f1c3e0ad57f1f1de5f311  7.210816e+10           zero       1.25   
...                         ...           ...            ...        ...   
24844  666e080a465f309038ab17db  8.523916e+10           zero       4.00   
24845  666e080a465f309038ab17db  8.523916e+10           1.00       4.00   
24846  611d67bbc9c88754a8a85f3e  8.523916e+10  

In [160]:
products_transactions = products_clean.merge(
    transaction_clean,
    left_on='BARCODE_str',
    right_on='BARCODE_str',
    how='left'
)
products_transactions

CATEGORY_1                                CATEGORY_2  \
0       Health & Wellness                             Sexual Health   
1                  Snacks                             Puffed Snacks   
2       Health & Wellness                                 Hair Care   
3       Health & Wellness                                 Oral Care   
4       Health & Wellness                    Medicines & Treatments   
...                   ...                                       ...   
859804  Health & Wellness  Topical Muscle & Joint Relief Treatments   
859805             Snacks                                   Cookies   
859806             Snacks                                     Candy   
859807             Snacks                              Nuts & Seeds   
859808  Health & Wellness                                 First Aid   

                         CATEGORY_3 CATEGORY_4  \
0       Conductivity Gels & Lotions        NaN   
1              Cheese Curls & Puffs        NaN   
2             Hair Care Accessories        NaN   
3                        Toothpaste        NaN   
4                    Essential Oils        NaN   
...                             ...        ...   
859804               Braces & Wraps        NaN   
859805                          NaN        NaN   
859806             Confection Candy        NaN   
859807                    Hazelnuts        NaN   
859808               First Aid Kits        NaN   

                                             MANUFACTURER            BRAND  \
0                                                     NaN              NaN   
1                                                     NaN              NaN   
2                                PLACEHOLDER MANUFACTURER          ELECSOP   
3                                       COLGATE-PALMOLIVE          COLGATE   
4       MAPLE HOLISTICS AND HONEYDEW PRODUCTS INTERCHA...  MAPLE HOLISTICS   
...                                                   ...              ...   
859804                                                NaN              NaN   
859805                              TREEHOUSE FOODS, INC.        LOFTHOUSE   
859806                                HARIBO GMBH & CO KG           HARIBO   
859807                                     DOUBLE-COLA CO            JUMBO   
859808                                                 3M          NEXCARE   

           BARCODE_x     BARCODE_str RECEIPT_ID PURCHASE_DATE SCAN_DATE  \
0       7.964944e+11  796494407820.0        NaN           NaN       NaN   
1       2.327801e+10   23278011028.0        NaN           NaN       NaN   
2       4.618178e+11  461817824225.0        NaN           NaN       NaN   
3       3.500047e+10   35000466815.0        NaN           NaN       NaN   
4       8.068109e+11  806810850459.0        NaN           NaN       NaN   
...              ...             ...        ...           ...       ...   
859804  7.223016e+11  722301569399.0        NaN           NaN       NaN   
859805  4.182082e+10   41820818468.0        NaN           NaN       NaN   
859806  1.001672e+11  100167154940.0        NaN           NaN       NaN   
859807  7.539076e+10   75390755960.0        NaN           NaN       NaN   
859808  7.967933e+11  796793337781.0        NaN           NaN       NaN   

       STORE_NAME USER_ID  BARCODE_y FINAL_QUANTITY FINAL_SALE  \
0             NaN     NaN        NaN            NaN        NaN   
1             NaN     NaN        NaN            NaN        NaN   
2             NaN     NaN        NaN            NaN        NaN   
3             NaN     NaN        NaN            NaN        NaN   
4             NaN     NaN        NaN            NaN        NaN   
...           ...     ...        ...            ...        ...   
859804        NaN     NaN        NaN            NaN        NaN   
859805        NaN     NaN        NaN            NaN        NaN   
859806        NaN     NaN        NaN            NaN        NaN   
859807        NaN     NaN        NaN            NaN        NaN 

In [140]:
len(products_transactions), len(products_clean), len(transaction_clean)


(859809, 841527, 49968)

CONCLUSIONS:
* Matching products and transactions (one to many) only 24849 out of +800K match
* Matching users and transactions (one to many) only 262 out of +100K match